### ISSUE TITLE
Eachine E016H support

### ISSUE THREAD

**Issue by BLeibbrand** [goebish/nrf24_multipro#46]
Hi all,
Thanks for the wonderfull work.
I love flying with my Taranis X9D+ and IRX4+ TX Module in combination with an Eachine 010.

I bougth an Eachine E016H which looks similar to the E010 but with more prop protection. 
I have kids and dogs so protection is needed.
But sofar I'm not ably to get the combination working.
Which protocol do need to use, or do you have to reverse engineer it first?

OpenTX Companion - 2.2.3
Taranis  FW:opentx-x9d+ version 2.2.3 (ae5ac05b)
IRX4+ - original (no flashing yet. Need to get USB - FTDI connector)

Eachine E016H
https://www.banggood.com/Eachine-E016H-Mini-Altitude-Hold-Headless-Mode-8mins-Flight-Time-2_4G-RC-Drone-quadcopter-RTF-p-1379560.html
![eachine e016h](https://user-images.githubusercontent.com/47399360/52377184-67ea1100-2a64-11e9-97e5-36f7e4871f7d.png)

**Comment by goebish** [comment]
If it doesn't work with any (sub)protocols with DeviationTX or MultiModule then it needs to be reverse engineered. Can you open the transmitter and quad and tell me which ICs they're using ?

**Comment by BLeibbrand** [comment]
Hello Goebish,
 
Thanks for the quick response.
I did some home work.
I hope this helps.
 
Yours Sincerely, 
 
 
Bert Leibbrand                                      <><
M +31 (0)6 27 74 60 88
 <mailto:bert.leibbrand@itri.nl> bert.leibbrand@itri.nl 
 
 
Met vriendelijke groet, 
 
 
Bert Leibbrand                                      <>< 
M +31 (0)6 27 74 60 88
 <mailto:bert.leibbrand@itri.nl> bert.leibbrand@itri.nl
 
Van: goebish [mailto:notifications@github.com] 
Verzonden: woensdag 6 februari 2019 23:16
Aan: goebish/nrf24_multipro
CC: BLeibbrand; Author
Onderwerp: Re: [goebish/nrf24_multipro] Eachine E016H support (#46)
 
If it doesn't work with any (sub)protocols with DeviationTX or MultiModule then it needs to be reverse engineered. Can you open the transmitter and quad and tell me which ICs they're using ?

**Comment by goebish** [comment]
I see nothing. If you want to post pictures I suppose you've to use github, you can't insert them in an email.

**Comment by BLeibbrand** [comment]
[PAN163CX  Productspecificatie.docx](https://github.com/goebish/nrf24_multipro/files/2839702/PAN163CX.Productspecificatie.docx)
[PAN163CX  Productspecificatie.pdf](https://github.com/goebish/nrf24_multipro/files/2839703/PAN163CX.Productspecificatie.pdf)
[SchematicsforESP32.pdf](https://github.com/goebish/nrf24_multipro/files/2839704/SchematicsforESP32.pdf)
![20181222_002009](https://user-images.githubusercontent.com/47399360/52396058-9db4e700-2ab0-11e9-9e34-4e4f432202bf.jpg)
![20181216_222324](https://user-images.githubusercontent.com/47399360/52396059-9e4d7d80-2ab0-11e9-8311-de876483980c.jpg)

**Comment by BLeibbrand** [comment]
![top view](https://user-images.githubusercontent.com/47399360/52396905-c094ca80-2ab3-11e9-995c-91405d6bd293.png)
![Uploading bottom view.png…]()

**Comment by goebish** [comment]
Ok, that's a PAN163CX, it can be reverse engineered, but not with a logic analyzer.
If someone dropships one to me I can take a look.

**Comment by BLeibbrand** [comment]
Hello Goebish,
I just had time to unscrew the transmitter.
The transmitter pictures.
The chip: XNS1042CV
![20190209_143407](https://user-images.githubusercontent.com/47399360/52521523-95a7a380-2c78-11e9-8efb-ce44409ddb6c.jpg)
![20190209_142847](https://user-images.githubusercontent.com/47399360/52521524-96403a00-2c78-11e9-80fa-328ce70405ba.jpg)
![20190209_142858](https://user-images.githubusercontent.com/47399360/52521525-96403a00-2c78-11e9-8b00-821b2c9c020e.jpg)
![20190209_143104](https://user-images.githubusercontent.com/47399360/52521526-96403a00-2c78-11e9-9cf8-f698c099dae2.jpg)
![20190209_143232](https://user-images.githubusercontent.com/47399360/52521528-96d8d080-2c78-11e9-9dac-c7c070fc8b2f.jpg)
extra info about transmitters with the XNS chip:
https://www.google.com/imgres?imgurl=http%3A%2F%2Fimg.over-blog-kiwi.com%2F0%2F55%2F43%2F29%2F20181129%2Fob_463a88_xwqz238505.jpg&imgrefurl=http%3A%2F%2Fchezcyril.over-blog.com%2F2018%2F11%2Fmini-drone-pliant-volcano-w606-16-un-drone-qui-se-pilote-grace-a-un-gant-et-au-mouvement-de-la-main-gyroscope.html&docid=ztPRrdg6XvaYsM&tbnid=nXG7m4h00eZ-eM%3A&vet=10ahUKEwjnhpD3467gAhWCPFAKHdPHDGEQMwgoKAAwAA..i&w=1205&h=544&bih=642&biw=1396&q=XNS%201042CV&ved=0ahUKEwjnhpD3467gAhWCPFAKHdPHDGEQMwgoKAAwAA&iact=mrc&uact=8

http://chezcyril.over-blog.com/2018/11/mini-drone-pliant-volcano-w606-16-un-drone-qui-se-pilote-grace-a-un-gant-et-au-mouvement-de-la-main-gyroscope.html

**Comment by goebish** [comment]
As I said, I believe it can be reverse engineered but I don't have one. If someone wants to dropship one to me send me an email @gmail.com

**Comment by goebish** [comment]
https://www.youtube.com/watch?v=GXEfK9D_90Y
😜

**Comment by BLeibbrand** [comment]
Great !!
Which steps do I have to take to get it into my Taranis / iRange?
 
Met vriendelijke groet, 
 
 
Bert Leibbrand                                      <>< 
M +31 (0)6 27 74 60 88
 <mailto:bert.leibbrand@itri.nl> bert.leibbrand@itri.nl
 
Van: goebish [mailto:notifications@github.com] 
Verzonden: woensdag 27 februari 2019 23:05
Aan: goebish/nrf24_multipro
CC: BLeibbrand; Author
Onderwerp: Re: [goebish/nrf24_multipro] Eachine E016H support (#46)
 
https://www.youtube.com/watch?v=GXEfK9D_90Y
😜

**Comment by goebish** [comment]
I've to cleanup my code then I'll contact @pascallanger so he can add it to the 4-in-1 module firmware.

**Comment by BLeibbrand** [comment]
What do you think of the indoor capabilities?
 
Met vriendelijke groet, 
 
 
Bert Leibbrand                                      <>< 
M +31 (0)6 27 74 60 88
 <mailto:bert.leibbrand@itri.nl> bert.leibbrand@itri.nl
 
Van: goebish [mailto:notifications@github.com] 
Verzonden: woensdag 27 februari 2019 23:17
Aan: goebish/nrf24_multipro
CC: BLeibbrand; Author
Onderwerp: Re: [goebish/nrf24_multipro] Eachine E016H support (#46)
 
I've to cleanup my code then I'll contact @pascallanger <https://github.com/pascallanger>  so he can add it to the 4-in-1 module firmware.

**Comment by goebish** [comment]
It's fun but quickly boring ;)
It should be perfect for a beginner or a kid as throttle is easy to manage.

**Comment by BLeibbrand** [comment]
Did you also enable the three fly modes?
 
Met vriendelijke groet, 
 
 
Bert Leibbrand                                      <>< 
M +31 (0)6 27 74 60 88
 <mailto:bert.leibbrand@itri.nl> bert.leibbrand@itri.nl
 
Van: goebish [mailto:notifications@github.com] 
Verzonden: woensdag 27 februari 2019 23:36
Aan: goebish/nrf24_multipro
CC: BLeibbrand; Author
Onderwerp: Re: [goebish/nrf24_multipro] Eachine E016H support (#46)
 
It's fun but quickly boring ;)
It should be perfect for a beginner or a kid as throttle is easy to manage.

**Comment by goebish** [comment]
Fly mode is always set to "expert" then you can set lower endpoints on your transmitter if required.

**Comment by pascallanger** [comment]
New Multiprotocol version 1.2.1.26 available on GitHub: https://github.com/pascallanger/DIY-Multiprotocol-TX-Module

Thanks to Goebish the E016H is now supported:
Protocol E01X (45)
Subprotocol E016H (2)
Channels:
CH5=STOP
CH6=FLIP
CH8=HEADLESS
CH9=RTH
Please test and report.

Pascal

**Comment by goebish** [comment]
Thank you!
@BLeibbrand please report once you've tested.

**Comment by goebish** [comment]
@pascallanger I think there's a mistake here:
https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/blob/04ed613ca3bd4bcda53457504adeca7aa9053168/Multiprotocol/E01X_nrf24l01.ino#L269
it should be XN297_SetTXAddr for E016H and HS6200_SetTXAddr for E012 & E015

**Comment by pascallanger** [comment]
Thanks! copy and paste issue... Corrected in v1.2.1.27.

**Comment by BLeibbrand** [comment]
Hello Goebish and Pascal,
 
I have soldered the connectors in my 4in1 and have to wait untill Monday to pickup my USB-TTL to get the bootloader installed.
As soon as I have tested it I will inform you.
 
Yours Sincerely, 
 
 
Bert Leibbrand                                      <><
M +31 (0)6 27 74 60 88
 <mailto:bert.leibbrand@itri.nl> bert.leibbrand@itri.nl 
 
Van: goebish [mailto:notifications@github.com] 
Verzonden: donderdag 28 februari 2019 21:20
Aan: goebish/nrf24_multipro
CC: BLeibbrand; Mention
Onderwerp: Re: [goebish/nrf24_multipro] Eachine E016H support (#46)
 
Thank you!
@BLeibbrand <https://github.com/BLeibbrand>  please report once you've tested.

**Comment by BLeibbrand** [comment]
The ST-LINK V2 arrived today. No luck in connecting to the 4in1.
Any help for the cable connection would be fine.
![image](https://user-images.githubusercontent.com/47399360/53738173-c7361800-3e8e-11e9-8bc5-17c7e7ef95d4.png)
And 4in1 GND, TX, RX

**Comment by pascallanger** [comment]
Why a st-link? you need a FTDI like the one in the documentation: https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/blob/master/docs/Compiling_STM32.md#tools-required
Have you tried to plug the USB to your PC? The latest modules are shipped with the bootloader flashed.

**Comment by goebish** [comment]
The STLink connects to a SWD interface (swclk, swdio, gnd), if you want connect to UART (gnd, rx, tx) you need an USB to UART adapter (FTDI dongle) instead.
I'm not a 4-in-1 expert, you'd better ask on rcgroups https://www.rcgroups.com/forums/showthread.php?2165676-DIY-Multiprotocol-TX-Module/

**Comment by BLeibbrand** [comment]
I already tried the USB: 
![image](https://user-images.githubusercontent.com/47399360/53746827-526cd900-3ea2-11e9-9a49-e7845c1ef3f6.png) 
It's a 4in1 from june 2018.
So I'll get a USB-UART and after that we can continu.

**Comment by pascallanger** [comment]
Yep no bootloader installed. Make sure to purchase the right FTDI which can run at 3.3v as mentionned on the github doc.

**Comment by BLeibbrand** [comment]
Hello guys, The FTDI finally arrived. 
I have loaded the version with the bootloader. 
The USB is now recognized as com25.  
OpenTX is version Companion is 2.2.3 (7 jan 2019)
![image](https://user-images.githubusercontent.com/47399360/55354098-22c2e800-54c5-11e9-90f1-6db105844839.png)
I can only see the first two sub protocols.
![image](https://user-images.githubusercontent.com/47399360/55354198-5aca2b00-54c5-11e9-9ba8-e48cbec696c0.png)
In the Taranis:
Mode:                     MULT   E01X E012 or E015 (no E016)
Module Status        V1.2.1.29
What is missing / what should I do?

**Comment by goebish** [comment]
Seems OpenTX Companion doesn't allow the selection of a custom protocol #
On the transmitter, select Custom protocol, with options 45 & 2:
![multi](https://user-images.githubusercontent.com/1297665/55355438-94506580-54c8-11e9-98a1-eccc651689a2.png)

**Comment by BLeibbrand** [comment]
It still doesn't bind

**Comment by goebish** [comment]
line 268 is wrong too:
https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/blob/04ed613ca3bd4bcda53457504adeca7aa9053168/Multiprotocol/E01X_nrf24l01.ino#L268
it should be:
if(sub_protocol!=E016H)

**Comment by BLeibbrand** [comment]
![image](https://user-images.githubusercontent.com/47399360/55358454-b8fc0b80-54cf-11e9-93c2-a3e4d4ea300d.png)
Still only blinking drone.
I'm using Custom 45 & 2 
Bind and bind on power up

**Comment by BLeibbrand** [comment]
I hope you have a next step to try.

**Comment by goebish** [comment]
Sorry, revert the change on line 268, I wasn't looking at the last version of the file, it should be
if (sub_protocol == E016H)

**Comment by goebish** [comment]
Your last error in the Arduino IDE is normal

**Comment by BLeibbrand** [comment]
OK. Now it does bind.
If you give it about 65% throttle the rotors spin (smilar to the spin with the standard transmitter just after binding with the throttle UP and DOWN). (After some time it does sometimes take of but I'm sure why).
SF will disarm
SB will blink 2x

**Comment by goebish** [comment]
I don't understand your last message, does it fly ?

**Comment by BLeibbrand** [comment]
Left and Right seem to be opposite.
Yes, it does fly as soon as you move the throttle a bit down and up again like 80% to 70% and up again.
The spinning only starts above the 60-70% below that no motors.

**Comment by goebish** [comment]
Ok that's normal.
You can revert Aileron and/or Rudder direction in the mixer.

**Comment by BLeibbrand** [comment]
So the drone is flying. 
Attitude hold is good.
Drifting to the right. (trim complete to the left to make it stable)
Headless and Stop are working.
RTH seems to fly to the right.
No space for flip at this moment

**Comment by goebish** [comment]
if it drifts put it on a level surface then set both sticks to bottom corners (I don't remember which ones) to recalibrate the accelerometer.
Or set a slight offset in the mixer for this channel.

**Comment by pascallanger** [comment]
E016 should be fixed. Can you download the latest version from github and try?

**Comment by BLeibbrand** [comment]
![image](https://user-images.githubusercontent.com/47399360/55435688-ac90b500-559a-11e9-8bc6-1815268410b2.png)
After reversing and trimming the drones is stable.

**Comment by goebish** [comment]
Ok, I'll check if calibration requires an extra flag.
Is one key take off really required as pushing throttle just does the same ?

**Comment by BLeibbrand** [comment]
One Key Take off would be fine to give new users an easy way to get the drone up without the up-down a bit - up  sequence needed at this moment. 
So I would say: Nice to have.

Calibration seems to be needed as the trim and the strange RTH direction might be caused by it.

**Comment by BLeibbrand** [comment]
Can you see if this drone also supports other flight modes (like acro) because as you mentioned before. If you are able to fly a drone this might becomes boring when you have some flying space.

**Comment by goebish** [comment]
One key take off would be a mess to implement, and not every controller has a momentary switch, honestly that's not a big deal if you've to push throttle instead of a button (this quad is *very* gentle ;))
Don't expect hidden features (acro mode...) on this kind of quad.
I found the calibration flag, I'm implementing it.

**Comment by goebish** [comment]
@pascallanger added calibration flag:
https://github.com/DeviationTX/deviation/pull/844/files

**Comment by BLeibbrand** [comment]
I have uploaded the newer version (1.2.1.30);
It is not recognized in the Taranis:
Mode: MULT E01X E012 or E015 (no E016)
So I still use the Custom 45 & 2
I have removed the reverses as they are fixed.
I only have to glue some components...
The rest is for tomorrow.

**Comment by goebish** [comment]
That's normal, E016 will only appear in the menu with the next OpenTX update.

**Comment by BLeibbrand** [comment]
Funny to see my other Eachine (E010 mini) is in an other group MJQX instead of E01X.
![image](https://user-images.githubusercontent.com/47399360/55439816-08147000-55a6-11e9-9c6c-5c2909b860cc.png)

**Comment by pascallanger** [comment]
Calibration available in the new version.

**Comment by BLeibbrand** [comment]
V 1.2.1.31
BIND:
* Power on drone'
* Power on Taranis [Bind on powerup = checked] 
No binding. Menu => page => [bind] => binding.
(Can I use an other method for binding with this drone?)

Calibration:
After binding, both sticks to bottom outside corner: drone blinking about 5 times.
So calibration seems to work.

The weather was OK today, so I could also test the Flip.
If I have the flip-switch "On" the drone will flip even with a little stick movement.
Can I switch it On and Off before giving the flip command to ensure it will flip only once?

Thanks sofar.

**Comment by pascallanger** [comment]
You need to learn to use your multi module and radio ;-)
 - multi offers an autobind option which if you have left the defaults in config.h enables you to bind with a switch assigned to channel 16. So basically, enable autobind in the model config and add a switch (temporary trainer switch is perfect) to channel 16. Each time you press it, it will launch a bind. Documentation: https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/blob/master/Protocols_Details.md#useful-notes-and-definitions
 - On your Taranis, you can configure a logical switch that will just send a pulse to the flip channel.
Pascal

**Comment by BLeibbrand** [comment]
If found some info about three speed levels.

It is the first button in the manual (top left).
Is this already included?

https://drones-camera.com/eachine-e016h-mini-drone-review-in-depth/

------------------
Ease Of Use:
When it comes to drone use, it is pretty simple to control. Its control is fundamental that allows beginners to get hold of it very quickly. It has low battery alarm as well that gives you an indication about when to bring it back. **Multiple speed modes bring more comfort to the flight**. Most of the drones need smartphone app download to control, but in this one, things are pretty simple, you can control it with the controller only.
-----------------

**Comment by goebish** [comment]
The protocol always use highest speed, if you need to tame it down just set lower rates (80%, 60%) in the mixer for aileron / elevator channel, as you would do with any hobby grade model.

**Comment by Dgirdler** [comment]
Gentlemen.  I have the Eachine E016H and the Taranis X9D+.  Can you please direct me to where I can learn how to put V 1.2.1.31 (or the newest V) into this...Multiprotocol TX Module For Frsky X9D.  (https://www.banggood.com/Multiprotocol-TX-Module-For-Frsky-X9D-X9D-Plus-X12S-Flysky-TH9X-Transmitter-p-1097688.html?rmmds=search&cur_warehouse=CN)

**Comment by goebish** [comment]
This project is outdated and I never supported this banggood module, if you want the E016H protocol, get a 4-in-1 Multi-Protocole Module.
https://github.com/pascallanger/DIY-Multiprotocol-TX-Module

**Comment by Dgirdler** [comment]
Thank you....learning....

Vr,
Dave

**Comment by derFrickler** [comment]
Any info if the new Eachine E016F uses the same protocol?
https://www.banggood.com/Eachine-E016F-3-in-1-EPP-Flying-Air-Boat-Land-Driving-Mode-Detachable-One-Key-Return-RC-Quadcopter-RTF-p-1472961.html

**Comment by Arkango** [comment]
Hello i found in a drone remote controller, an xns1042cv can be this discussion appropriate? That component implements the E016H protocol ?

**Comment by goebish** [comment]
xns1042cv is a 8 bit mcu with built-in xn297 rf transceiver.
The E016H transmitter is using this IC but that doesn't mean that any other machine that's using the same chip is running the same protocol.

**Comment by Arkango** [comment]
@goebish Ok how can I understand the protocol ?

**Comment by little-Beppo** [comment]
Hello, I'm new here.
I've already dealt with the topic of multi-protocols a little and now my question.
I bought the current E016H, but I cannot bind it with the DIY-Multiprotocol V1.3.1.69. I found the chip "FC298TX GS50XFN411" in the transmitter.
is this the wrong version or the wrong protocol?
Please help me .
Thank you

**Comment by goebish** [comment]
Seems like this is a new version, the original version I reverse engineered was using XNSxxxx / PANxxxx chips, I've never heard of yours before :/

**Comment by little-Beppo** [comment]
that is the picture from the transmitter chip.
![P1001317-sm](https://user-images.githubusercontent.com/74399343/99067404-f3ce8c80-25aa-11eb-8ebb-9d0a4f094bd1.jpg)
so, is there a possibility to get them in the air?

**Comment by little-Beppo** [comment]
let me know if you need more informations

**Comment by goebish** [comment]
There's not much I can do.
Your best bet is asking on the multi protocol thread for some guidance on dumping xn297 packets, but this is a unknown chip, no idea if it's xn297 compatible or not.
https://www.rcgroups.com/forums/showthread.php?2165676-DIY-Multiprotocol-TX-Module

**Comment by little-Beppo** [comment]
OK anyway, many thanks for your support.

**Comment by diogo-f** [comment]
I can't bind with jumper t18 and 5in1 updated with latest version multi-t18int-aetr-v1.3.1.77 as it is now available, tried with reta and v1.3.1.69 too.

**Comment by goebish** [comment]
@diogo-f seems like there's a new version of this quad which is not compatible with the current x-in-1 protocol implementation.
Open the transmitter and check the chip inside, if it's XNSxxxx then it should work, if it's FC298TX  it's not supported.

**Comment by diogo-f** [comment]
Yes it says FC298RX, is it possible to add support for this new version?

**Comment by goebish** [comment]
Please read above, I suspect those chips are rebranded (XNSxxxx ?), perhaps they're xn297 compatible.
Ask for guidance about dumping XN297 packets on the forum, an external module is a better choice for that since it requires to be connected to the computer (usb). It can be done with an internal (5-in-1) module but does require to connect a UART to USB adapter internally ...

**Comment by goebish** [comment]
@pascallanger or maybe you've already asked someone else to try to dump the packets and it didn't work ? (not xn297 compatible ...)

**Comment by little-Beppo** [comment]
Hello,
I created the dumps with the help of Pascallanger.
https://github.com/pascallanger/DIY-Multiprotocol-TX-Module/issues/469
He is of the opinion that the new E016H is not compatible with the Xn297.
He want test a new one.

**Comment by goebish** [comment]
thanks @little-Beppo, I'm closing this issue.
